gpt_engineer/steps.py

```diff
@@ -380,6 +395,72 @@ def fix_code(ai: AI, dbs: DBs):
     return messages
 
 
+def self_heal(ai: AI, dbs: DBs):
+    """Attempts to execute the code from the entrypoint and if it fails,
+    sends the error output back to the AI with instructions to fix.
+
+    This code will make `MAX_SELF_HEAL_ATTEMPTS` to try and fix the code
+    before giving up.
+
+    This makes the assuption that the previous step was `gen_entrypoint`,
+    this code could work with `simple_gen`, or `gen_clarified_code` as well.
+    """
+
+    # step 1. execute the entrypoint
+    log_path = dbs.workspace.path / "log.txt"
+
+    attempts = 0
+    messages = []
+
+    while attempts < MAX_SELF_HEAL_ATTEMPTS:
+        log_file = open(log_path, "w")  # wipe clean on every iteration
+        timed_out = False
+
+        p = subprocess.Popen(  # attempt to run the entrypoint
+            "bash run.sh",
+            shell=True,
+            cwd=dbs.workspace.path,
+            stdout=log_file,
+            stderr=log_file,
+            bufsize=0,
+        )
+        try:  # timeout if the process actually runs
+            p.wait(timeout=ASSUME_WORKING_TIMEOUT)
+        except subprocess.TimeoutExpired:
+            timed_out = True
+            print("The process hit a timeout before exiting.")
+
+        # get the result and output
+        # step 2. if the return code not 0, package and send to the AI
+        if p.returncode != 0 and not timed_out:
+            print("run.sh failed.  Let's fix it.")
+
+            # pack results in an AI prompt
```

Comment:
So this PR doesnt actually doself heal right its just the framework in place? is the idea to combine this with the new improve code?

Reply:
No its does self-heal, but it rewrites the whole program in every iteration. I see it as a first pass.
The advantage with having an incremental self-heal using --improve would be that it eats less time and token.

Reply:
Nice- whats the plan for merging? I'd suggest if shall we just get the conflicts resolved and merge it without adding to the default steps for now? Then we can continue to improve in main branch

Reply:
Im happy to do this today if you agree

Reply:
I will work on getting the conflicts resolved.